stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp

```diff
@@ -112,126 +118,67 @@ typename return_type<T_loc, T_cut>::type ordered_logistic_lpmf(
- * the number of cutpoints plus 2; if the cutpoint vector is
- * empty; if the cutpoint vector contains a non-positive,
- * non-finite value; or if the cutpoint vector is not sorted in
- * ascending order.
- * @throw std::invalid_argument If y and lambda are different
- * lengths, or if y and the array of cutpoints are of different
- * lengths.
- */
-template <bool propto, typename T_loc, typename T_cut>
-typename return_type<T_loc, T_cut>::type ordered_logistic_lpmf(
-    const std::vector<int>& y,
-    const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
-    const std::vector<Eigen::Matrix<T_cut, Eigen::Dynamic, 1> >& c) {
-  typename return_type<T_loc, T_cut>::type logp_n(0.0);
+  check_consistent_sizes(function, "Integers", y, "Locations", lambda);
 
-  using std::exp;
-  using std::log;
+  for (int n = 0; n < N; n++) {
+    check_bounded(function, "Random variable", y_vec[n], 1, K);
+    check_finite(function, "Location parameter", lam_vec[n]);
+  }
 
-  static const char* function = "ordered_logistic";
+  for (size_t i = 0, size_ = length(c_vec); i < size_; i++) {
+    check_ordered(function, "Cut-points", c_vec[i]);
+    check_greater(function, "Size of cut points parameter", c_vec[i].size(), 0);
+    check_finite(function, "Final cut-point", c_vec[i](c_vec[i].size() - 1));
+    check_finite(function, "First cut-point", c_vec[i](0));
+  }
 
-  int N = lambda.size();
+  ordLog_helper<T_partials_return> ordhelp;
+  operands_and_partials<T_loc, T_cut> ops_partials(lambda, c);
 
-  check_consistent_sizes(function, "Integers", y, "Locations", lambda);
-  check_consistent_sizes(function, "Integers", y, "Cut-points", c);
+  T_partials_return logp(0.0);
 
-  for (int i = 0; i < N; ++i) {
-    int K = c[i].size() + 1;
+  for (int n = 0; n < N; ++n) {
+    for (size_t i = 0, size_ = length(c_vec); i < size_; i++) {
+      T_partials_return lam_dbl = value_of(lam_vec[n]);
+      T_partials_vec c_dbl
+          = value_of(c_vec[i]).template cast<T_partials_return>();
+      T_partials_vec d = ordhelp.deriv(y_vec[n], K, lam_dbl, c_dbl);
 
-    check_bounded(function, "Random variable", y[i], 1, K);
-    check_greater(function, "Size of cut points parameter", c[i].size(), 0);
-    check_ordered(function, "Cut-points", c[i]);
-  }
+      logp += ordhelp.logp(y_vec[n], K, lam_dbl, c_dbl);
 
-  check_finite(function, "Location parameter", lambda);
-  check_finite(function, "Cut-points", c);
+      if (!is_constant_struct<T_loc>::value)
+        ops_partials.edge1_.partials_[n] = d[0];
 
-  for (int i = 0; i < N; ++i) {
-    int K = c[i].size() + 1;
-    if (y[i] == 1) {
-      logp_n -= log1p_exp(lambda[i] - c[i][0]);
-    } else if (y[i] == K) {
-      logp_n -= log1p_exp(c[i][K - 2] - lambda[i]);
-    } else {
-      logp_n += log_inv_logit_diff(lambda[i] - c[i][y[i] - 2],
-                                   lambda[i] - c[i][y[i] - 1]);
+      if (!is_constant_struct<T_cut>::value)
+        ops_partials.edge2_.partials_vec_[n] += d.tail(K - 1);
```

Comment:
Can you migrate the ordhelp stuff back into this function? The issue is that this isn't gonna be efficient.

There's only going to be 2-3 non-zero elements of d every time, so doing these vector adds is just adding a lot of zeros.

Also since there aren't multiple functions now and these things only get called once, there isn't a huge gain to be had by separating these things out.

Reply:
The difficulty with removing the vector adds is that the number of non-zero elements will differ based on the value of ```y```, so it would require moving the ``` if (!is_constant_struct``` statements into the if statements above. I'm not sure which of the two would be more efficient (or if there's a third option that would be even better).

Reply:
There's no cost to the tempate metaprogram other than in compile time---it'll all get neatly unfolded with optimization.  Conditioning on whether something is zero before adding can save some memory, but it can be slower.  What are you suggesting to do here, @bbbales2?

Reply:
If there's structure to where the zeroes are, you could use that.

Reply:
@andrjohns It'll be more efficient with the change. +1 to what Bob said about the condition on type

The issue is that a vector of length K - 1 gets accumulated N times, but in reality there are only somewhere between 2 or 3 nonzeros in each of the N accumulations.

Reply:
I see---it's the same `d` and `K` each time!  For `n >> 1` (the case we care most about for efficiency) and for `K >> 2`, computing the nonzeros first, then looping to add them should be more efficient.

Reply:
This is ready for review again, the upstream failures *seem* like jenkins just having a bad day